src/Core/Framework/DataAbstractionLayer/Dbal/EntityDefinitionQueryHelper.php

```diff
@@ -449,50 +448,37 @@ public function addTranslationSelect(string $root, EntityDefinition $definition,
             return;
         }
 
-        $fields = $translationDefinition->getFields();
+        $fields = $translationDefinition->getFields()->filter(
+            fn (Field $field) => $field instanceof StorageAware
+                && $definition->getFields()->get($field->getPropertyName()) instanceof TranslatedField,
+        );
         if (!empty($partial)) {
-            $fields = $translationDefinition->getFields()->filter(fn (Field $field) => $field->is(PrimaryKey::class)
-                || isset($partial[$field->getPropertyName()])
-                || $field instanceof FkField);
+            $fields = $fields->filter(fn (Field $field) => isset($partial[$field->getPropertyName()]));
```

Comment:
why did you removed the primary key and fk field here. seems not related to your changes?

Reply:
Because they were never read.

They just used to end up in the outer-most `SELECT` when fetching the entities for hydration but were never accessed. Just like `TranslationFieldResolver` this method used to select more fields than necessary, so I removed them.

Here's are example queries I just dumped using the following criteria (`fields` to apply the `$partial` filter):
```
{
    "fields": [
        "name"
    ],
    "filter": [
        {
            "type": "contains",
            "field": "name",
            "value": "a"
        }
    ]
}
```
Old query:
```sql
SELECT
    `product`.`id` AS `product.id`,
    `product`.`version_id` AS `product.versionId`,
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
    `product.translation.fallback_1.name`,
    `product.parent.translation.fallback_1.name`,
    `product.translation.name`,
    `product.parent.translation.name`,
    COALESCE(
        `product.translation.fallback_1.name`,
        `product.parent.translation.fallback_1.name`,
        `product.translation.name`,
        `product.parent.translation.name`,
        `product.translation.name`
    ) AS `product.name`,
    `product.translation.fallback_1.productId`,
    `product.parent.translation.fallback_1.productId`,
    `product.translation.productId`,
    `product.parent.translation.productId`,
    `product.translation.fallback_1.languageId`,
    `product.parent.translation.fallback_1.languageId`,
    `product.translation.languageId`,
    `product.parent.translation.languageId`,
    `product.translation.fallback_1.productVersionId`,
    `product.parent.translation.fallback_1.productVersionId`,
    `product.translation.productVersionId`,
    `product.parent.translation.productVersionId`
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
FROM `product`
LEFT JOIN `product` `product.parent`
    ON `product`.`parent_id` = `product.parent`.`id`
    AND `product`.`parent_version_id` = `product.parent`.`version_id`
LEFT JOIN (
    SELECT
        `product.translation`.product_id,
        `product.translation`.product_version_id,
        `product.translation`.meta_description AS `product.translation.metaDescription`,
        `product.translation`.name AS `product.translation.name`,
        `product.translation`.keywords AS `product.translation.keywords`,
        `product.translation`.description AS `product.translation.description`,
        `product.translation`.meta_title AS `product.translation.metaTitle`,
        `product.translation`.pack_unit AS `product.translation.packUnit`,
        `product.translation`.pack_unit_plural AS `product.translation.packUnitPlural`,
        `product.translation`.custom_search_keywords AS `product.translation.customSearchKeywords`,
        `product.translation`.slot_config AS `product.translation.slotConfig`,
        `product.translation`.custom_fields AS `product.translation.customFields`,
        `product.translation`.created_at AS `product.translation.createdAt`,
        `product.translation`.updated_at AS `product.translation.updatedAt`,
        `product.translation`.product_id AS `product.translation.productId`,
        `product.translation`.language_id AS `product.translation.languageId`,
        `product.translation`.product_version_id AS `product.translation.productVersionId`,
        `product.translation.fallback_1`.meta_description AS `product.translation.fallback_1.metaDescription`,
        `product.translation.fallback_1`.name AS `product.translation.fallback_1.name`,
        `product.translation.fallback_1`.keywords AS `product.translation.fallback_1.keywords`,
        `product.translation.fallback_1`.description AS `product.translation.fallback_1.description`,
        `product.translation.fallback_1`.meta_title AS `product.translation.fallback_1.metaTitle`,
        `product.translation.fallback_1`.pack_unit AS `product.translation.fallback_1.packUnit`,
        `product.translation.fallback_1`.pack_unit_plural AS `product.translation.fallback_1.packUnitPlural`,
        `product.translation.fallback_1`.custom_search_keywords AS `product.translation.fallback_1.customSearchKeywords`,
        `product.translation.fallback_1`.slot_config AS `product.translation.fallback_1.slotConfig`,
        `product.translation.fallback_1`.custom_fields AS `product.translation.fallback_1.customFields`,
        `product.translation.fallback_1`.created_at AS `product.translation.fallback_1.createdAt`,
        `product.translation.fallback_1`.updated_at AS `product.translation.fallback_1.updatedAt`,
        `product.translation.fallback_1`.product_id AS `product.translation.fallback_1.productId`,
        `product.translation.fallback_1`.language_id AS `product.translation.fallback_1.languageId`,
        `product.translation.fallback_1`.product_version_id AS `product.translation.fallback_1.productVersionId`
    FROM `product_translation` `product.translation`
    LEFT JOIN `product_translation` `product.translation.fallback_1`
        ON `product.translation`.`product_id` = `product.translation.fallback_1`.`product_id`
        AND `product.translation.fallback_1`.language_id = :languageId
        AND `product.translation`.product_version_id = `product.translation.fallback_1`.product_version_id
    WHERE
        `product.translation`.language_id = :languageId
) `product.product_translation`
    ON `product.product_translation`.`product_id` = `product`.`id`
    AND `product.product_translation`.product_version_id = `product`.version_id
LEFT JOIN (
    SELECT
        `product.parent.translation`.product_id,
        `product.parent.translation`.product_version_id,
        `product.parent.translation`.meta_description AS `product.parent.translation.metaDescription`,
        `product.parent.translation`.name AS `product.parent.translation.name`,
        `product.parent.translation`.keywords AS `product.parent.translation.keywords`,
        `product.parent.translation`.description AS `product.parent.translation.description`,
        `product.parent.translation`.meta_title AS `product.parent.translation.metaTitle`,
        `product.parent.translation`.pack_unit AS `product.parent.translation.packUnit`,
        `product.parent.translation`.pack_unit_plural AS `product.parent.translation.packUnitPlural`,
        `product.parent.translation`.custom_search_keywords AS `product.parent.translation.customSearchKeywords`,
        `product.parent.translation`.slot_config AS `product.parent.translation.slotConfig`,
        `product.parent.translation`.custom_fields AS `product.parent.translation.customFields`,
        `product.parent.translation`.created_at AS `product.parent.translation.createdAt`,
        `product.parent.translation`.updated_at AS `product.parent.translation.updatedAt`,
        `product.parent.translation`.product_id AS `product.parent.translation.productId`,
        `product.parent.translation`.language_id AS `product.parent.translation.languageId`,
        `product.parent.translation`.product_version_id AS `product.parent.translation.productVersionId`,
        `product.parent.translation.fallback_1`.meta_description AS `product.parent.translation.fallback_1.metaDescription`,
        `product.parent.translation.fallback_1`.name AS `product.parent.translation.fallback_1.name`,
        `product.parent.translation.fallback_1`.keywords AS `product.parent.translation.fallback_1.keywords`,
        `product.parent.translation.fallback_1`.description AS `product.parent.translation.fallback_1.description`,
        `product.parent.translation.fallback_1`.meta_title AS `product.parent.translation.fallback_1.metaTitle`,
        `product.parent.translation.fallback_1`.pack_unit AS `product.parent.translation.fallback_1.packUnit`,
        `product.parent.translation.fallback_1`.pack_unit_plural AS `product.parent.translation.fallback_1.packUnitPlural`,
        `product.parent.translation.fallback_1`.custom_search_keywords AS `product.parent.translation.fallback_1.customSearchKeywords`,
        `product.parent.translation.fallback_1`.slot_config AS `product.parent.translation.fallback_1.slotConfig`,
        `product.parent.translation.fallback_1`.custom_fields AS `product.parent.translation.fallback_1.customFields`,
        `product.parent.translation.fallback_1`.created_at AS `product.parent.translation.fallback_1.createdAt`,
        `product.parent.translation.fallback_1`.updated_at AS `product.parent.translation.fallback_1.updatedAt`,
        `product.parent.translation.fallback_1`.product_id AS `product.parent.translation.fallback_1.productId`,
        `product.parent.translation.fallback_1`.language_id AS `product.parent.translation.fallback_1.languageId`,
        `product.parent.translation.fallback_1`.product_version_id AS `product.parent.translation.fallback_1.productVersionId`
    FROM `product_translation` `product.parent.translation`
    LEFT JOIN `product_translation` `product.parent.translation.fallback_1`
        ON `product.parent.translation`.`product_id` = `product.parent.translation.fallback_1`.`product_id`
        AND `product.parent.translation.fallback_1`.language_id = :languageId
        AND `product.parent.translation`.product_version_id = `product.parent.translation.fallback_1`.product_version_id
    WHERE
        `product.parent.translation`.language_id = :languageId
) `product.product_translation.parent`
    ON `product.product_translation.parent`.`product_id` = `product.parent`.`id`
    AND `product.product_translation.parent`.product_version_id = `product.parent`.version_id
WHERE
    `product`.`version_id` = :versionId
```

New query:
```sql
SELECT
    `product`.`id` AS `product.id`,
    `product`.`version_id` AS `product.versionId`,
    `product.translation.override_1.name`,
    `product.parent.translation.override_1.name`,
    `product.translation.name`,
    `product.parent.translation.name`,
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
    COALESCE(
        `product.translation.override_1.name`,
        `product.parent.translation.override_1.name`,
        `product.translation.name`,
        `product.parent.translation.name`
    ) AS `product.name`
    # [END] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
FROM `product`
LEFT JOIN `product` `product.parent`
    ON `product`.`parent_id` = `product.parent`.`id`
    AND `product`.`parent_version_id` = `product.parent`.`version_id`
LEFT JOIN (
    SELECT
        `product.translation`.`product_id`,
        `product.translation`.`product_version_id`,
        `product.translation`.`name` AS `product.translation.name`,
        `product.translation.override_1`.`name` AS `product.translation.override_1.name`
    FROM `product_translation` `product.translation`
    LEFT JOIN `product_translation` `product.translation.override_1`
        ON `product.translation`.`product_id` = `product.translation.override_1`.`product_id`
        AND `product.translation.override_1`.`language_id` = :languageId
        AND `product.translation`.`product_version_id` = `product.translation.override_1`.`product_version_id`
    WHERE
        `product.translation`.`language_id` = :languageId
) `product.product_translation`
    ON `product.product_translation`.`product_id` = `product`.`id`
    AND `product.product_translation`.`product_version_id` = `product`.`version_id`
LEFT JOIN (
    SELECT
        `product.parent.translation`.`product_id`,
        `product.parent.translation`.`product_version_id`,
        `product.parent.translation`.`name` AS `product.parent.translation.name`,
        `product.parent.translation.override_1`.`name` AS `product.parent.translation.override_1.name`
    FROM `product_translation` `product.parent.translation`
    LEFT JOIN `product_translation` `product.parent.translation.override_1`
        ON `product.parent.translation`.`product_id` = `product.parent.translation.override_1`.`product_id`
        AND `product.parent.translation.override_1`.`language_id` = :languageId
        AND `product.parent.translation`.`product_version_id` = `product.parent.translation.override_1`.`product_version_id`
    WHERE
        `product.parent.translation`.`language_id` = :languageId
) `product.parent.product_translation`
    ON `product.parent.product_translation`.`product_id` = `product.parent`.`id`
    AND `product.parent.product_translation`.`product_version_id` = `product.parent`.`version_id`
WHERE
    `product`.`version_id` = :versionId
```